Regr automation migration and enhancement for replace loaded worker nodes with larger instance types.

svetsa-rh opened the pull request:
See [OCPQE-11961](https://issues.redhat.com//browse/OCPQE-11961) and [OCPQE-12152](https://issues.redhat.com//browse/OCPQE-12152)

https://issues.redhat.com/browse/OCPQE-11961
https://issues.redhat.com/browse/OCPQE-12152

Test case: https://polarion.engineering.redhat.com/polarion/redirect/project/OSE/workitem?id=OCP-40557

Also added parameters to specify cloud type and number of replicas to create when creating new machineset.

Here are couple of successful runs:
https://mastern-jenkins-csb-openshift-qe.apps.ocp-c1.prod.psi.redhat.com/job/scale-ci/job/e2e-benchmarking-multibranch-pipeline/job/regression-test/67/console
https://mastern-jenkins-csb-openshift-qe.apps.ocp-c1.prod.psi.redhat.com/job/scale-ci/job/e2e-benchmarking-multibranch-pipeline/job/regression-test/65/console


@paigerube14 and @qiliRedHat PTAL

paigerube14 commented on perfscale_regression_ci/scripts/cluster_density_env.sh:
```diff
@@ -0,0 +1,8 @@
+export E2E_BENCHMARKING_REPOSITORY=${E2E_BENCHMARKING_REPOSITORY:-"https://github.com/cloud-bulldozer/e2e-benchmarking"}
+export E2E_BENCHMARKING_BRANCH=${E2E_BENCHMARKING_BRANCH:-"master"}
+# Common ENVs for kube-burner 
+export WORKLOAD=${WORKLOAD:-"cluster-density"}
```
let's just set this WORKLOAD parameter in the tests env.sh file. If we set a different workload than custom it won't get overwritten

paigerube14 commented on perfscale_regression_ci/scripts/scalability/replace_nodes/clouds/alicloud.sh:
```diff
@@ -0,0 +1,20 @@
+export CLUSTER_NAME=$(oc get machineset -n openshift-machine-api -o=go-template='{{(index (index .items 0).metadata.labels "machine.openshift.io/cluster-api-cluster" )}}')
```
Let's create a more common env.sh file for items like this that are consistent among all cloud types, this way if we ever need to edit we can in one place

paigerube14 commented on perfscale_regression_ci/scripts/scalability/replace_nodes/clouds/gcp.sh:
```diff
@@ -0,0 +1,22 @@
+export GCP_SERVICE_ACCOUNT_EMAIL=openshift-qe.iam.gserviceaccount.com
+
+export PWD=$(pwd)
+export PYTHONPATH=${PWD}/../../:../../../utils/:../../../:${PYTHONPATH}
```
what is this doing?

paigerube14 commented on perfscale_regression_ci/scripts/scalability/replace_nodes/replacenodes.py:
```diff
@@ -0,0 +1,49 @@
+import sys
+import os
+import datetime
+from utils.ocp_utils import run, get_machine_replicas, scale_machine_replicas, wait_for_node_deletion, wait_for_node_creation, cluster_health_check, delete_machineset
```
here you could just do **import utils.ocp_utils as ocp_utils** and then call the function by ocp_utils.<function_name>
It would look a little cleaner and don't have to write out all the function names

paigerube14 commented on perfscale_regression_ci/utils/cleanup.py:
```diff
@@ -0,0 +1,49 @@
+import json
+
+# Invokes a given command and returns the stdout
+def invoke(command):
```
let's use the run in the utils/ocp_utils to keep consistent because it does pretty much the same thing

paigerube14 commented on perfscale_regression_ci/utils/ocp_utils.py:
```diff
@@ -0,0 +1,126 @@
+
+import subprocess
+import time 
+#import sys
```
if these packages aren't being used, let's remove them

paigerube14 commented on perfscale_regression_ci/utils/ocp_utils.sh:
```diff
@@ -0,0 +1,26 @@
+
+#set -ex
+
+cluster_health_check() {
```
let's pick either this function or the one in the ocp_utils.py but not leave both since they are duplications of each other

paigerube14 commented on perfscale_regression_ci/utils/ocp_utils.sh:
```diff
@@ -0,0 +1,26 @@
+    set +x
+}
+
+install_dittybopper() {
```
This could also be moved to [common.sh here](https://github.com/openshift/svt/blob/master/perfscale_regression_ci/scripts/common.sh)

svetsa-rh commented:
Closing this PR as the test strategy and refactoring of some of the functions here are needed.

svetsa-rh commented:
@paigerube14 PTAL - I have updated and/or refactored few things along with test strategy.
https://mastern-jenkins-csb-openshift-qe.apps.ocp-c1.prod.psi.redhat.com/job/scale-ci/job/e2e-benchmarking-multibranch-pipeline/job/regression-test/78/ is a passing test run.

paigerube14 commented on perfscale_regression_ci/scripts/scalability/replace_loadedworker_nodes.sh:
```diff
@@ -0,0 +1,75 @@
+    echo "Deploy new machineset and scale down one machine at a time from existing machinesets starting at: $(date) " 
+    cd ./replace_nodes/clouds
+    . ./${CLOUD}.sh
+    cd ../..
```
Are you missing a run_workload here? I don't see it before you scale the nodes

paigerube14 replied:
Or is this a changed step?

svetsa-rh commented:
@paigerube14 and @qiliRedHat Can you PTAL? I have updated to use custom_workload_env.sh and also incorporated other changes suggested by Paige.

Pre-requisites for running test case is documented here:
https://github.com/svetsa-rh/svt/blob/476652ddd49b988eb9d16836a910b53cf2d35fff/perfscale_regression_ci/scripts/scalability/replace_loadedworker_nodes.sh#L8

Here is the latest successful run: 
https://mastern-jenkins-csb-openshift-qe.apps.ocp-c1.prod.psi.redhat.com/job/scale-ci/job/e2e-benchmarking-multibranch-pipeline/job/regression-test/88/console

qiliRedHat commented on perfscale_regression_ci/scripts/scalability/replace_loadedworker_nodes.sh:
```diff
@@ -0,0 +1,73 @@
+### 1) Prior to running this script, make sure to run loaded-upgrade Jenkins job 
+### to create a new cluster with one of the below profiles and run cluster-density workload
+### https://mastern-jenkins-csb-openshift-qe.apps.ocp-c1.prod.psi.redhat.com/job/scale-ci/job/e2e-benchmarking-multibranch-pipeline/job/loaded-upgrade/
+### Cluster with 3 master nodes and 30 worker nodes using one of these profiles:
```
Cluster with 3 m5.2xlarge master nodes and 30 m5.xlarge worker cluster on AWS, or equivalent size of these profiles:

qiliRedHat commented on perfscale_regression_ci/scripts/scalability/replace_nodes/clouds/aws.sh:
```diff
@@ -0,0 +1,11 @@
+source ./common.sh
+
+export OPENSHIFT_INFRA_NODE_VOLUME_SIZE=128
```
I didn't see INFRA related ENV vars used anywhere, but it's fine if you think we will use it in the future.

qiliRedHat commented on perfscale_regression_ci/scripts/scalability/replace_loadedworker_nodes.sh:
```diff
@@ -0,0 +1,73 @@
+export JOB_ITERATIONS="${params[2]}"
+
+export WORKLOAD="cluster-density"
+echo "Testing ${WORKLOAD} on $CLOUD using $REPLICAS replicas for $JOB_ITERATIONS job iterations for new machineset."
```
This test is not actually run in code, right?, it is run in [Pre-requisites](https://github.com/svetsa-rh/svt/blob/476652ddd49b988eb9d16836a910b53cf2d35fff/perfscale_regression_ci/scripts/scalability/replace_loadedworker_nodes.sh#L8). So let's remove this output to avoid confusion.

qiliRedHat commented on perfscale_regression_ci/scripts/scalability/utils/cleanup.py:
```diff
@@ -0,0 +1,37 @@
+    print("job type " + str(job))
+    if job.lower() == "pod-density":
+        job = "node-density"
+    ocp_utils.run("oc delete ns --wait=false -l kube-burner-job=" + job)
```
This python file did cleanup for kube-burner specific resources. I suggest to change the name cleanup.py to a more specific one, like cleanup_kubeburner.py. 
A 'nice to have' suggestion, I think this can be moved to the [common place](https://github.com/openshift/svt/tree/master/perfscale_regression_ci/scripts)  so test cases under other sub-folders (network/performance) can also make use of it.

paigerube14 commented:
@svetsa-rh I think you need to rebase with master. Don't think you updated the kubeburner object templates, correct?

svetsa-rh commented:
@paigerube14 and @qiliRedHat - rebased with master and made all necessary changes we discussed. Can you PTAL?

Passing Jenkins Run:
https://mastern-jenkins-csb-openshift-qe.apps.ocp-c1.prod.psi.redhat.com/job/scale-ci/job/svetsa-e2e-benchmarking-multibranch-pipeline/job/regression-test/8/console

qiliRedHat commented:
/lgtm

qiliRedHat commented:
@paigerube14 Let me know if you have more comments.

qiliRedHat commented:
@svetsa-rh I recalled in our last discussion before I went to holiday, we wanted to add an output to tell user to get dittybopper screenshot in 5 minutes.

svetsa-rh commented:
Hi @qiliRedHat I removed that since we are now generating csv files with the necessary data (before and after nodes are replaced) and hence doesn't require manual intervention. I still left code to install dittybopper in case tester wants to see the charts at any given point but it is not required to verify the results.

paigerube14 commented on perfscale_regression_ci/scripts/scalability/utils/ocp_utils.py:
```diff
@@ -0,0 +1,124 @@
+        else:
+            break
+
+if(sys.argv[1]=='cluster_health_check'):
```
do you need this line? I see you're calling cluster_health_check directly from replacenodes

paigerube14 commented on perfscale_regression_ci/scripts/scalability/utils/cleanup.py:
```diff
@@ -0,0 +1,37 @@
+    ocp_utils.run("oc delete ns --wait=false -l kube-burner-job=" + job)
+    wait_for_all_deleted_ns(job)
+
+def wait_for_all_deleted_ns(job, wait_num=300):
```
this is a duplicate function of: https://github.com/openshift/svt/blob/7364b2b76b8ddc23fd74ca11b3c6c9f81ddeddcb/perfscale_regression_ci/scripts/common.sh#L25

paigerube14 commented on perfscale_regression_ci/scripts/scalability/utils/cleanup.py:
```diff
@@ -0,0 +1,37 @@
+import sys
+import ocp_utils 
+
+def delete_all_namespaces(job):
```
same as below, duplicated function here: https://github.com/openshift/svt/blob/7364b2b76b8ddc23fd74ca11b3c6c9f81ddeddcb/perfscale_regression_ci/scripts/common.sh#L45

paigerube14 commented on perfscale_regression_ci/scripts/scalability/replace_loadedworker_nodes_env.sh:
```diff
@@ -0,0 +1,8 @@
+export INDEXING=true
+export METADATA_COLLECTION=true
+export METRICS_PROFILE="metrics-profiles/metrics-aggregated.yaml"
+export COMPARISON_CONFIG="clusterVersion.json nodeMasters-max.json nodeAggWorkers.json"
```
COMPARISON_CONFIG and GEN_CSV should be added to this file to be used be all the workloads https://github.com/openshift/svt/blob/master/perfscale_regression_ci/scripts/custom_workload_env.sh

svetsa-rh commented:
@paigerube14 Can you PTAL?
New run with changes
https://mastern-jenkins-csb-openshift-qe.apps.ocp-c1.prod.psi.redhat.com/job/scale-ci/job/svetsa-e2e-benchmarking-multibranch-pipeline/job/regression-test/13/console

paigerube14 commented on perfscale_regression_ci/scripts/scalability/replace_loadedworker_nodes.sh:
```diff
@@ -0,0 +1,69 @@
+if [ $? -eq 0 ]; 
+then 
+    # Cluster health check prior to testing
+    python ./utils/ocp_utils.py cluster_health_check
```
Let's do something like [this line](https://github.com/openshift/svt/blob/7364b2b76b8ddc23fd74ca11b3c6c9f81ddeddcb/openshift_scalability/scripts/pod_density/pod_density.sh#L62) so we can take out  (if(sys.argv[1]=='cluster_health_check'):) in ocp_utils.py and just call the function directly

paigerube14 commented on perfscale_regression_ci/scripts/scalability/replace_loadedworker_nodes.sh:
```diff
@@ -0,0 +1,69 @@
+    python ./replacenodes.py ${CLOUD} ${REPLICAS} ${OPENSHIFT_WORKER_NODE_INSTANCE_TYPE}
+    echo "Existing machines scaled down and new nodes are up."
+    sleep 180 
+    python ./utils/ocp_utils.py cluster_health_check
```
You are already doing a cluster health check at the end of the replacenodes script, can we take this one out

svetsa-rh commented:
@paigerube14 Updated. Can you PTAL?
Here is a run in progress but shows cluster health check with changes.
https://mastern-jenkins-csb-openshift-qe.apps.ocp-c1.prod.psi.redhat.com/job/scale-ci/job/svetsa-e2e-benchmarking-multibranch-pipeline/job/regression-test/14/console

svetsa-rh commented:
@qiliRedHat Can you ptal. Not sure if you need to mention lgtm label again. Since I pushed more changes label is removed.

qiliRedHat commented:
/lgtm
